cmd/restic/cmd_mount.go

```diff
@@ -0,0 +1,301 @@
+// +build !openbsd
+
+package main
+
+import (
+	"encoding/binary"
+	"fmt"
+	"os"
+	"time"
+
+	"golang.org/x/net/context"
+
+	"github.com/restic/restic"
+	"github.com/restic/restic/backend"
+	"github.com/restic/restic/crypto"
+	"github.com/restic/restic/pack"
+	"github.com/restic/restic/repository"
+
+	"bazil.org/fuse"
+	"bazil.org/fuse/fs"
+)
+
+type CmdMount struct {
+	global *GlobalOptions
+	ready  chan struct{}
+}
+
+func init() {
+	_, err := parser.AddCommand("mount",
+		"mount a repository",
+		"The mount command mounts a repository read-only to a given directory",
+		&CmdMount{
+			global: &globalOpts,
+			ready:  make(chan struct{}, 1),
+		})
+	if err != nil {
+		panic(err)
+	}
+}
+
+func (cmd CmdMount) Usage() string {
+	return "MOUNTPOINT"
+}
+
+func (cmd CmdMount) Execute(args []string) error {
+	if len(args) == 0 {
+		return fmt.Errorf("wrong number of parameters, Usage: %s", cmd.Usage())
+	}
+
+	repo, err := cmd.global.OpenRepository()
+	if err != nil {
+		return err
+	}
+
+	err = repo.LoadIndex()
+	if err != nil {
+		return err
+	}
+
+	mountpoint := args[0]
+	if _, err := os.Stat(mountpoint); err != nil {
+		if os.IsNotExist(err) {
+			cmd.global.Verbosef("Mountpoint %s doesn't exist, creating it\n", mountpoint)
+			err = os.Mkdir(mountpoint, os.ModeDir|0700)
+			if err != nil {
+				return err
+			}
+		}
+	}
+	c, err := fuse.Mount(
+		mountpoint,
+		fuse.ReadOnly(),
+		fuse.FSName("restic"),
+	)
+	if err != nil {
+		return err
+	}
+
+	root := fs.Tree{}
+	root.Add("snapshots", &snapshots{
+		repo:           repo,
+		knownSnapshots: make(map[string]*restic.Snapshot),
+	})
+
+	cmd.global.Printf("Now serving %s under %s\n", repo.Backend().Location(), mountpoint)
+	cmd.global.Printf("Don't forget to umount after quitting !\n")
+
+	cmd.ready <- struct{}{}
+
+	err = fs.Serve(c, &root)
+	if err != nil {
+		return err
+	}
+
+	<-c.Ready
+	return c.MountError
+}
+
+// These lines statically ensure that a *snapshots implement the given
+// interfaces; a misplaced refactoring of the implementation that breaks
+// the interface will be catched by the compiler
+var _ = fs.HandleReadDirAller(&snapshots{})
+var _ = fs.NodeStringLookuper(&snapshots{})
+
+type snapshots struct {
+	repo *repository.Repository
+
+	// snapshot timestamp -> snapshot
+	knownSnapshots map[string]*restic.Snapshot
```

Comment:
I don't know how the go fuse package you are using is implemented in terms of concurrency, but do we need to mutex the access to this map so that it can safely be used concurrently? for example if ReadDirAll or Lookup are called concurrently?